### PR TITLE
feat: allow tasks to be indented / unintended within the task list

### PR DESCRIPTION
**Pull request by jacobboertjes**
Addressing feature request #6 by allowing tasks to be indented up to 3 times. User is also able to unindent them.

### CODE REVIEW

**Comment by jacobboertjes** [comment]
image


image

**Comment by jacobboertjes** [comment]
@Toofifty @cecilia-sanare Could I get some eyes on this?

**Comment by cecilia-sanare** [comment]
Weird, GitHub just *didn't* notify me about this PR.
Checking it out now!

File: src/main/java/com/toofifty/goaltracker/ui/TaskItemContent.java [hunk 1]
```diff
@@ -42,7 +42,10 @@ public void refresh()
     {
         titleLabel.setText(task.toString());
         titleLabel.setForeground(STATUS_TO_COLOR.get(task.getStatus()));
+
+        int indent = 16 * task.getSubtaskRank();
         iconLabel.setIcon(iconService.get(task));
+        iconLabel.setBorder(new EmptyBorder(4, indent, 0, 0));
```
Comment (cecilia-sanare) [inline src/main/java/com/toofifty/goaltracker/ui/TaskItemContent.java:48]: Mind removing the top padding from this?

```suggestion
        iconLabel.setBorder(new EmptyBorder(0, indent, 0, 0));
```

File: src/main/java/com/toofifty/goaltracker/models/task/Task.java [hunk 2]
```diff
@@ -21,10 +22,36 @@ public abstract class Task
     @SerializedName("has_been_notified")
     private boolean notified = false;
 
+    @Builder.Default
+    @SerializedName("subtask_rank")
+    private int subtask = 0;
```
Comment (cecilia-sanare) [inline src/main/java/com/toofifty/goaltracker/models/task/Task.java:27]: Would you mind renaming this to something like `indentLevel`?

File: src/main/java/com/toofifty/goaltracker/models/task/Task.java [hunk 3]
```diff
@@ -21,10 +22,36 @@ public abstract class Task
...
+
     public boolean isDone() {
         return Status.COMPLETED.equals(this.status);
     }
 
+    public int getSubtaskRank() {
+        return subtask;
+    }
+
+    public void setSubtask()
```
Comment (cecilia-sanare) [inline src/main/java/com/toofifty/goaltracker/models/task/Task.java:39]: Mind renaming this to `indent`?

File: src/main/java/com/toofifty/goaltracker/models/task/Task.java [hunk 4]
```diff
@@ -21,10 +22,36 @@ public abstract class Task
...
+    {
+        if (subtask < MAX_SUBTASK)
+        {
+            subtask += 1;
+        }
+    }
+
+    public void unsetSubtask()
```
Comment (cecilia-sanare) [inline src/main/java/com/toofifty/goaltracker/models/task/Task.java:47]: Mind renaming this to `dedent` or `unindent` either is fine?

File: src/main/java/com/toofifty/goaltracker/utils/ReorderableList.java [hunk 5]
```diff
@@ -33,6 +33,10 @@ public void moveToBottom(T item)
         this.add(item);
     }
 
+    public void indent(T item)
```
Comment (cecilia-sanare) [inline src/main/java/com/toofifty/goaltracker/utils/ReorderableList.java:36]: Unused method

File: src/main/java/com/toofifty/goaltracker/ui/components/ListItemPanel.java [hunk 6]
```diff
@@ -146,4 +172,16 @@ public void onRemoved(Consumer<T> removeListener) {
     public void onReordered(Consumer<T> reorderListener) {
         this.reorderedListener = reorderListener;
     }
+
+    public ListItemPanel<T> onIndented(Consumer<T> indentedListener) {
```
Comment (cecilia-sanare) [inline src/main/java/com/toofifty/goaltracker/ui/components/ListItemPanel.java:176]:
```suggestion
    public void onIndented(Consumer<T> indentedListener) {
```

File: src/main/java/com/toofifty/goaltracker/ui/components/ListItemPanel.java [hunk 7]
```diff
@@ -146,4 +172,16 @@ public void onRemoved(Consumer<T> removeListener) {
...
+        this.indentedListener = indentedListener;
+
+        return this;
```
Comment (cecilia-sanare) [inline src/main/java/com/toofifty/goaltracker/ui/components/ListItemPanel.java:179]:
```suggestion
```

File: src/main/java/com/toofifty/goaltracker/ui/components/ListItemPanel.java [hunk 8]
```diff
@@ -146,4 +172,16 @@ public void onRemoved(Consumer<T> removeListener) {
...
+    }
+
+    public ListItemPanel<T> onUnindented(Consumer<T> unindentedListener) {
```
Comment (cecilia-sanare) [inline src/main/java/com/toofifty/goaltracker/ui/components/ListItemPanel.java:182]:
```suggestion
    public void onUnindented(Consumer<T> unindentedListener) {
```

File: src/main/java/com/toofifty/goaltracker/ui/components/ListItemPanel.java [hunk 9]
```diff
@@ -146,4 +172,16 @@ public void onRemoved(Consumer<T> removeListener) {
...
+        this.unindentedListener = unindentedListener;
+
+        return this;
```
Comment (cecilia-sanare) [inline src/main/java/com/toofifty/goaltracker/ui/components/ListItemPanel.java:185]:
```suggestion
```

File: src/main/java/com/toofifty/goaltracker/ui/components/ListItemPanel.java [hunk 10]
```diff
@@ -21,12 +24,17 @@ public class ListItemPanel<T> extends JPanel implements Refreshable
...
+    private final JMenuItem indentTask = new JMenuItem("Indent");
+    private final JMenuItem unindentTask = new JMenuItem("Unindent");
     private final JPopupMenu popupMenu = new JPopupMenu();
 
     private final ReorderableList<T> list;
     private final T item;
+
     private Consumer<T> reorderedListener;
     private Consumer<T> removedListener;
+    private Consumer<T> indentedListener = null;
```
Comment (cecilia-sanare) [inline src/main/java/com/toofifty/goaltracker/ui/components/ListItemPanel.java:36]:
```suggestion
    private Consumer<T> indentedListener;
```

File: src/main/java/com/toofifty/goaltracker/ui/components/ListItemPanel.java [hunk 11]
```diff
@@ -12,6 +12,9 @@
 import javax.swing.JPanel;
 import javax.swing.JPopupMenu;
 import javax.swing.border.EmptyBorder;
+
+import org.apache.commons.lang3.ObjectUtils.Null;
```
Comment (cecilia-sanare) [inline src/main/java/com/toofifty/goaltracker/ui/components/ListItemPanel.java:16]:
```suggestion
```

File: src/main/java/com/toofifty/goaltracker/models/task/Task.java [hunk 12]
```diff
@@ -21,10 +22,36 @@ public abstract class Task
...
+    @SerializedName("subtask_rank")
+    private int subtask = 0;
+
+    transient final private int MAX_SUBTASK = 3;
+
     public boolean isDone() {
         return Status.COMPLETED.equals(this.status);
     }
 
+    public int getSubtaskRank() {
```
Comment (cecilia-sanare) [inline src/main/java/com/toofifty/goaltracker/models/task/Task.java:35]: Mind removing this and putting the `@Getter` notation on the property?

File: src/main/java/com/toofifty/goaltracker/ui/components/ListItemPanel.java [hunk 13]
```diff
@@ -104,6 +122,14 @@ public void refreshMenu()
...
         }
         popupMenu.add(removeTask);
+        if (this.indentedListener != null)
+        {
+            popupMenu.add(indentTask);
+        }
+        if (this.unindentedListener != null)
+        {
+            popupMenu.add(unindentTask);
+        }
```
Comment (cecilia-sanare) [inline src/main/java/com/toofifty/goaltracker/ui/components/ListItemPanel.java:132]: Mind moving these into the `!list.isFirst(item)` if statement?

**Review by cecilia-sanare — CHANGES_REQUESTED** [review]

**Comment by cecilia-sanare** [comment]
Overall it looks really good! Mostly just naming and a little bit of cleanup!
I super appreciate you taking the time to submit a PR! :heart:

File: src/main/java/com/toofifty/goaltracker/ui/components/ListItemPanel.java (hunk 13 above)
Comment (jacobboertjes) [reply src/main/java/com/toofifty/goaltracker/ui/components/ListItemPanel.java:132]: Smart!

**Review by cecilia-sanare — APPROVED** [review]
LGTM! Thanks again for your contribution!~

**Comment by jacobboertjes** [comment]
Thanks for the comments. Realised I hadn't cleaned up very well before submitting the PR lol.
I updated the commit with the suggestions. Although not sure what is going wrong, in VsCode I'm suddenly seeing an error trying to run GoalTackerPluginTest 
image

**Comment by cecilia-sanare** [comment]
@jacobboertjes Oh strange, I pulled down your code and it worked perfectly fine on my end o:

**Comment by jacobboertjes** [comment]
@cecilia-sanare  yeah figured it's some environmental issue. If it works for you I'm sure it's good

**Pull request merged**